Use zlint to check our CRLs

aarongable opened the pull request:
Update zlint to v3.5.0, which introduces scaffolding for running lints over CRLs.

Convert all of our existing CRL checks to structs which match the zlint interface, and add them to the registry. Then change our linter's CheckCRL function, and crl-checker's Validate function, to run all lints in the zlint registry.

This change touches a lot of files, but involves almost no logic changes. It's all just infrastructure, changing the way our lints and their tests are shaped, and moving test files into new homes.

~Do not merge until https://github.com/letsencrypt/boulder/pull/6977 is merged~
Fixes https://github.com/letsencrypt/boulder/issues/6934
Fixes https://github.com/letsencrypt/boulder/issues/6979

aarongable commented:
I strongly recommend reviewing this change one commit at a time, as in general each commit converts a single one of our old checks into the zlint format. This makes it much easier to confirm that the check logic has stayed the same in all cases.

jsha commented on linter/lints/rfc/lint_crl_has_issuer_name.go:
```diff
@@ -0,0 +1,64 @@
+package rfc
+
+/*
+ * ZLint Copyright 2023 Regents of the University of Michigan
```
How come this gained a copyright notice for the Regents of the University of Michigan?

If we're planning to contribute this lint upstream, wouldn't we add the copyright notice at the time we contribute it (assuming upstream has a relevant requirement about copyright notices)?

aarongable replied:
These were accidental, they get removed in the last commit in the stack.

jsha commented on linter/lints/rfc/lint_crl_has_valid_timestamps.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+func (l *crlHasValidTimestamps) Execute(c *x509.RevocationList) *lint.LintResult {
+	input := cryptobyte.String(c.RawTBSRevocationList)
```
It was really hard to review this diff to verify that all the substantive lines were moved and not changed. Note for other reviewers: here's how I did it:

```git show 55a8dc75b3ca872d161e80f97d12d32c5bd4d132:linter/lints/rfc/lint_crl_has_valid_timestamps.go > /tmp/new
git show 55a8dc75b3ca872d161e80f97d12d32c5bd4d132~:linter/lints/crl/lints.go > /tmp/old
vimdiff /tmp/{old,new}
```

I also manually edited `old` while diffing, to delete a lot of the head of lints.go, which helped the differ figure things out.

jsha commented on linter/lints/rfc/lint_crl_no_empty_revoked_certificates_list.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+func (l *crlNoEmptyRevokedCertsList) Execute(c *x509.RevocationList) *lint.LintResult {
+	if c.RevokedCertificates != nil && len(c.RevokedCertificates) == 0 {
```
Something not relevant to this PR, but I noticed passing through: this suggests a semantic difference between `nil` and "non-`nil` but 0 len`, which is not documented in crl_x509.

aarongable replied:
Good catch. I think the reason it's not documented is that to me, the person writing crl_x509, this was (in large part) _the whole point_. Being able to distinguish between "has no RevokedCertificates list" and "has an empty RevokedCertificates list" was one of the core things I wanted, and so it was just obvious to me that that's how it works, and so I completely forgot to document it.

All that said, this subtle distinction was (correctly, imo) rejected by the Go maintainers during the upstreaming process. So this lint will have to change when we unfork.

jsha commented:
For similar changes in the future, I think it would make them more reviewable to have a single commit that moves lines to their new files with zero changes (even if that doesn't compile), and then a followup change for each file that puts the file in the new style and makes it compile.

aarongable commented:
> For similar changes in the future, I think it would make them more reviewable to have a single commit that moves lines to their new files with zero changes (even if that doesn't compile), and then a followup change for each file that puts the file in the new style and makes it compile.

If you don't mind a force push, I can re-upload this PR with each of the commits split out in exactly this way. It'll just take me maybe half an hour to do the interactive rebase.

aarongable commented:
I've broken the zlint update out into https://github.com/letsencrypt/boulder/pull/6977 to make this one smaller.

aarongable commented:
Here's a draft PR with things even more broken out. Let me know what you think. I can make that the real PR, or force-push those commits here, if you prefer. https://github.com/letsencrypt/boulder/pull/6978

jsha commented:
> I've broken the zlint update out into #6977 to make this one smaller.

Nice. I've approved that one, though to be clear that zlint update didn't provide any difficulty in reviewing this, since it's its own commit.

> Here's a draft PR with things even more broken out. Let me know what you think. I can make that the real PR, or force-push those commits here, if you prefer. https://github.com/letsencrypt/boulder/pull/6978

Excellent, that's much better. Yes please to a force push to this PR.

aarongable commented:
> Yes please to a force push to this PR.

Done! I've also made a few small updates like incorporating a TODO addressing your comment on lint_crl_no_empty_revoked_certificates_list.go

jsha commented on crl/checker/checker_test.go:
```diff
@@ -39,9 +39,15 @@ func TestValidate(t *testing.T) {
+	test.AssertNotError(t, err, "parsing test crl")
 	err = Validate(crl, issuer, 100*365*24*time.Hour)
-	test.AssertError(t, err, "validaint crl with lint error")
+	test.AssertError(t, err, "validaing crl with lint error")
```
```suggestion
	test.AssertError(t, err, "validating crl with lint error")
```

jsha reviewed (changes requested):
Looks good, just one spelling fix to a spelling fix. :D

jsha commented on linter/lints/cabf_br/lint_crl_validity_period.go:
```diff
@@ -3,21 +3,53 @@ package cabfbr
+field MUST NOT be more than ten days beyond the value of the thisUpdate field.
+
+Although the validity period for CRLs covering the status of Subordinate CA
+certificates is longer (up to 12 months), Boulder does not produce such CRLs,
```
Technically Boulder (via cmd/ceremony) does produce Subordinate CA CRLs; I think we just don't lint them yet.

aarongable commented:
I've made one slight modification, updating the applicability date of one of these lints to not be a few days in the future from now. PTAL.

aarongable commented:
FWIW, that "Base automatically changed" message doesn't require re-review. That just means that previously this PR was "targeting" a non-`main` branch (specifically, the `update-zlint` branch from https://github.com/letsencrypt/boulder/pull/6977), and that branch got merged. No new commits or content were pushed to this branch, so the existing reviews can stand. But now that https://github.com/letsencrypt/boulder/pull/6981 has been merged into this branch, it should be given another look-over, so no we're in a good state anyway!

pgporada commented on cmd/crl-checker/main.go:
```diff
@@ -100,7 +100,14 @@ func main() {
+		zcrl, err := crl_x509.ParseRevocationList(crl.Raw)
+		if err != nil {
+			errCount += 1
+			logger.Errf("checking CRL %q failed: %s", u, err)
```
Nit: This log line is identical to line 113 which might be confusing if we ever had to debug this during an incident.
```suggestion
			logger.Errf("parsing CRL %q failed: %s", u, err)
```
